src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,183 @@
+                bool foundSnapshot = false;
+                foreach (var (currentPosition, currentSnapshot) in _incrementalSnapshots)
+                {
+                    if (currentPosition <= position)
+                    {
+                        incrementalSnapshot = currentSnapshot;
+                        foundSnapshot = true;
+                        break;
+                    }
+                }
+
+                if (!foundSnapshot)
+                {
+                    return null;
+                }
+
+                var globalState = _walkerGlobalStates[incrementalSnapshot.GlobalStateIndex];
+                var variableState = new VariableState(globalState.VariableSlot, globalState.VariableBySlot, globalState.VariableTypes, incrementalSnapshot.VariableState.Clone());
+                var method = globalState.Symbol as MethodSymbol;
+                return new NullableWalker(
+                    compilation,
+                    globalState.Symbol,
+                    useMethodSignatureParameterTypes: !(method is null),
+                    method,
+                    nodeToAnalyze,
+                    binder,
+                    binder.Conversions,
+                    variableState,
+                    returnTypesOpt: null,
+                    analyzedNullabilityMap,
+                    snapshotBuilderOpt: null);
+            }
+
+#if DEBUG
+            internal void VerifyNode(BoundNode node)
+            {
+                if (node.Kind == BoundKind.TypeExpression || node.WasCompilerGenerated)
+                {
+                    return;
+                }
+
+                Debug.Assert(_incrementalSnapshots.ContainsKey(node.Syntax.SpanStart), $"Did not find a snapshot for {node} `{node.Syntax}.`");
+                Debug.Assert(_walkerGlobalStates.Length > _incrementalSnapshots[node.Syntax.SpanStart].GlobalStateIndex, $"Did not find global state for {node} `{node.Syntax}`.");
+            }
+#endif
+
+            internal sealed class Builder
+            {
+                private readonly ArrayBuilder<SharedWalkerState> _walkerStates = ArrayBuilder<SharedWalkerState>.GetInstance();
+                private readonly ImmutableDictionary<int, Snapshot>.Builder _incrementalSnapshots = ImmutableDictionary.CreateBuilder<int, Snapshot>();
+                private readonly PooledDictionary<Symbol, int> _symbolToSlot = PooledDictionary<Symbol, int>.GetInstance();
+                private readonly ArrayBuilder<int> _previousWalkerSlots = ArrayBuilder<int>.GetInstance();
+                private int _currentWalkerSlot = -1;
+                private int _nextWalkerSlot = 0;
+
+                internal SnapshotManager ToManagerAndFree()
+                {
+                    Debug.Assert(_currentWalkerSlot == -1 && _previousWalkerSlots.Count == 0, "Attempting to finalize snapshots before all walks completed");
+                    Debug.Assert(_walkerStates.Count == _nextWalkerSlot);
```

Comment:
>Debug.Assert(_walkerStates.Count == _nextWalkerSlot); [](start = 20, length = 53)

Is there an advantage to have a separate field if we can always use ```_walkerStates.Count``` to get the value?